### ISSUE TITLE
ceiling(CRootOf()) not implemented, leads to exception in `Range`

### ISSUE THREAD

**Issue by gschintgen** [sympy/sympy#25965]
As the title says, `ceiling(CRootOf())` is not implemented. This can lead to issues if sets are involved. (E.g. when resolving trig equations.)
If I'm not mistaken it should be sufficient to add a special case to [`ceiling`'s `_eval_number()`](https://github.com/sympy/sympy/blob/2d2052cb56d8a675c7395717461833d44d683479/sympy/functions/elementary/integers.py#L304) method. (I may be wrong though...)

In the stracktrace below, `Range` complains about symbolic start/end due to an unevaluated `ceiling(CRootOf())`, even though `CRootOf` stands for a precisely defined real/complex number.

```Python
In [2]: var('n')
   ...: s = ImageSet(Lambda(n, n + (CRootOf(x**5 - x**2 + 1, 0))), Integers)
   ...: s.intersect(Interval(-10, 10))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[2], line 3
      1 var('n')
      2 s = ImageSet(Lambda(n, n + (CRootOf(x**5 - x**2 + 1, 0))), Integers)
----> 3 s.intersect(Interval(-10, 10))

File ~/sympy/sympy/sets/sets.py:150, in Set.intersect(self, other)
    131 def intersect(self, other):
    132     """
    133     Returns the intersection of 'self' and 'other'.
    134
   (...)
    148
    149     """
--> 150     return Intersection(self, other)

File ~/sympy/sympy/sets/sets.py:1507, in Intersection.__new__(cls, evaluate, *args)
   1505 if evaluate:
   1506     args = list(cls._new_args_filter(args))
-> 1507     return simplify_intersection(args)
   1509 args = list(ordered(args, Set._infimum_key))
   1511 obj = Basic.__new__(cls, *args)

File ~/sympy/sympy/sets/sets.py:2644, in simplify_intersection(args)
   2642 new_args = False
   2643 for t in args - {s}:
-> 2644     new_set = intersection_sets(s, t)
   2645     # This returns None if s does not know how to intersect
   2646     # with t. Returns the newly intersected set otherwise
   2648     if new_set is not None:

File ~/sympy/sympy/multipledispatch/dispatcher.py:243, in Dispatcher.__call__(self, *args, **kwargs)
    241     self._cache[types] = func
    242 try:
--> 243     return func(*args, **kwargs)
    245 except MDNotImplementedError:
    246     funcs = self.dispatch_iter(*types)

File ~/sympy/sympy/sets/handlers/intersection.py:399, in _(self, other)
    397     return S.EmptySet
    398 elif isinstance(range_set, Range) and range_set.size is not S.Infinity:
--> 399     range_set = FiniteSet(*list(range_set))
    401 if range_set is not None:
    402     return imageset(Lambda(n, f), range_set)

File ~/sympy/sympy/sets/fancysets.py:745, in Range.__iter__(self)
    743 n = self.size  # validate
    744 if not (n.has(S.Infinity) or n.has(S.NegativeInfinity) or n.is_Integer):
--> 745     raise TypeError("Cannot iterate over symbolic Range")
    746 if self.start in [S.NegativeInfinity, S.Infinity]:
    747     raise TypeError("Cannot iterate over Range with infinite start")

TypeError: Cannot iterate over symbolic Range
```

**Comment by oscarbenjamin** [comment]
The test here is not sufficient to guarantee iterability:
https://github.com/sympy/sympy/blob/2d2052cb56d8a675c7395717461833d44d683479/sympy/sets/handlers/intersection.py#L398-L399
Checking `range_set.is_iterable` would be better but still not fully enough.

**Comment by gschintgen** [comment]
There is also `utilities.iterables.iterable()` which is intended to check for finitely iterable lists, sets, etc., but it's not implemented yet/correctly for sets: #25992

**Comment by oscarbenjamin** [comment]
The fix for `ceiling(RootOf)` is:
```diff
diff --git a/sympy/functions/elementary/integers.py b/sympy/functions/elementary/integers.py
index fcb342a903..091f7c8d2f 100644
--- a/sympy/functions/elementary/integers.py
+++ b/sympy/functions/elementary/integers.py
@@ -49,10 +49,10 @@ def eval(cls, arg):
                 ipart += i*S.ImaginaryUnit
             elif (i := intof(t)) is not None:
                 ipart += i
-            elif t.has(Symbol):
-                spart += t
-            else:
+            elif t.is_number:
                 npart += t
+            else:
+                spart += t
 
         if not (npart or spart):
             return ipart
```
Then:
```python
In [1]: ceiling(RootOf(x**5-x**2+1,0))
Out[1]: 0

In [2]: s = ImageSet(Lambda(n, n + (CRootOf(x**5 - x**2 + 1, 0))), Integers)

In [3]: print(s.intersect(Interval(-10, 10)))
{-9 + CRootOf(x**5 - x**2 + 1, 0), -8 + CRootOf(x**5 - x**2 + 1, 0), -7 + CRootOf(x**5 - x**2 + 1, 0), -6 + CRootOf(x**5 - x**2 + 1, 0), -5 + CRootOf(x**5 - x**2 + 1, 0), -4 + CRootOf(x**5 - x**2 + 1, 0), -3 + CRootOf(x**5 - x**2 + 1, 0), -2 + CRootOf(x**5 - x**2 + 1, 0), -1 + CRootOf(x**5 - x**2 + 1, 0), CRootOf(x**5 - x**2 + 1, 0) + 1, CRootOf(x**5 - x**2 + 1, 0) + 2, CRootOf(x**5 - x**2 + 1, 0) + 3, CRootOf(x**5 - x**2 + 1, 0) + 4, CRootOf(x**5 - x**2 + 1, 0) + 5, CRootOf(x**5 - x**2 + 1, 0) + 6, CRootOf(x**5 - x**2 + 1, 0) + 7, CRootOf(x**5 - x**2 + 1, 0) + 8, CRootOf(x**5 - x**2 + 1, 0) + 9, CRootOf(x**5 - x**2 + 1, 0) + 10, CRootOf(x**5 - x**2 + 1, 0)}

In [4]: print(s.intersect(Interval(-10, 10)).evalf())
{-9.80873060047939, -8.80873060047939, -7.80873060047939, -6.80873060047939, -5.80873060047939, -4.80873060047939, -3.80873060047939, -2.80873060047939, -1.80873060047939, -0.808730600479392, 0.191269399520608, 1.19126939952061, 2.19126939952061, 3.19126939952061, 4.19126939952061, 5.19126939952061, 6.19126939952061, 7.19126939952061, 8.19126939952061, 9.19126939952061}
```
